### ISSUE TITLE
Unreachable code in HostingApplication.CreateContext(IFeatureCollection)

### ISSUE THREAD

**Issue by joacar** [aspnet/AspNetCore#17415]
### Describe the bug
The [if-statement](https://github.com/aspnet/AspNetCore/blob/16be9a264e48560e10a3ee9683ecaed342d4ca11/src/Hosting/Hosting/src/Internal/HostingApplication.cs#L62) is unreachable since direct cast never produce `null`-object.

The `httpContext.HttpContext` is **not** set if it's `DefaultHttpContext` but set in every other case. I just want to raise this, but it probably has its reason.

Propose
```c#
if (hostContext.HttpContext is DefaultHttpContext defaultHttpContext)
{
    _defaultHttpContextFactory.Initialize(defaultHttpContext, contextFeatures);
    httpContext = defaultHttpContext;
}
else
{
    httpContext = _defaultHttpContextFactory.Create(contextFeatures);
    hostContext.HttpContext = httpContext;    
}

// Current
var defaultHttpContext = (DefaultHttpContext)hostContext.HttpContext;
if (defaultHttpContext is null)
{
    httpContext = _defaultHttpContextFactory.Create(contextFeatures);
    hostContext.HttpContext = httpContext;
}
else
{
    _defaultHttpContextFactory.Initialize(defaultHttpContext, contextFeatures);
    httpContext = defaultHttpContext;
}
```

### To Reproduce
N/A

### Further technical details
- ASP.NET Core version **master**-branch
- ~Include the output of `dotnet --info`~
- ~The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version~

**Comment by Tratcher** [comment]
Regardless of casting, why can't the original hostContext.HttpContext be null? I'd expect it to be null on the first request, the DefaultHttpContext hasn't been created yet. After that it would be cached in the hostContext.

**Comment by jkotalik** [comment]
I think this can be unit tested as well to see if you get an InvalidCastException. Try writing a unit test where hostContext.HttpContext isn't a DefaultHttpContext.

**Comment by joacar** [comment]
I'll write the unit test and report my findings.

**Comment by anurse** [comment]
I'm pretty sure you *can* cast `null` values. It worked when I tried it briefly on sharplab.io ;).

**Comment by joacar** [comment]
@anurse Yes I've tried it in C# Interactive as well and that _ought_ to be close to reality :)

Lets consider the possibilities (the original code in mind)

1. The value is `null` and no exception is thrown and the `if`-branch is entered correctly.
2. The value is `DefaultHttpContext` and no exception is thrown and the `else`-branch is entered correctly.
3. The value is something else and an exception will be thrown

Scenario 3 is feasible, but probably never occurred (I guess someone should have fixed it by then :)), since `HttpContext` is public and not-sealed where as `DefaultHttpContext` is sealed. So any type _can_ be of type `HttpContext`, but not a sub-type of `DefaultHttpContext`.

So we're left with 1 and 2 which implies that the type **must** be `DefaultHttpContext`, however uninitialized.

As I'm writing this it becomes clear that my "problem" with the code is that, for me, it wasn't immediately clear that it must be of type `DefaultHttpContext`. 

Lesson learned - spend more time analyzing the code and think, think, think. Sorry for wasting your time.

_Note_ I made a small effort in getting this  code to compile, but was unsuccessful on both Windows 10 and WSL (Ubuntu 16.04) running the build.cmd/build.sh. If time I'll give it another try!
